src/plugins/git/porcelain.js

```diff
@@ -0,0 +1,264 @@
+    if (n.type() !== "COMMIT") {
+      throw new Error(`Unable to cast ${n.type()} to Commit`);
+    }
+    return new Commit(n.graph, n.nodeAddress);
+  }
+
+  hash(): Hash {
+    return this.address().id;
+  }
+
+  parents(): Commit[] {
+    return this.graph
+      .neighborhood(this.nodeAddress, {
+        nodeType: "COMMIT",
+        edgeType: "HAS_PARENT",
+        direction: "OUT",
+      })
+      .map(({neighbor}) => new Commit(this.graph, neighbor));
+  }
+
+  tree(): Tree {
+    const trees = this.graph
+      .neighborhood(this.nodeAddress, {
+        nodeType: "TREE",
+        edgeType: "HAS_TREE",
+        direction: "OUT",
+      })
+      .map(({neighbor}) => new Tree(this.graph, neighbor));
+    if (trees.length !== 1) {
+      throw new Error(
+        `Commit ${stringify(this.nodeAddress)} has wrong number of trees`
+      );
+    }
+    return trees[0];
+  }
+}
+
+export class Tree extends BaseGitPorcelain {
+  static from(n: BaseGitPorcelain): Tree {
+    if (n.type() !== "TREE") {
+      throw new Error(`Unable to cast ${n.type()} to Tree`);
+    }
+    return new Tree(n.graph, n.nodeAddress);
+  }
+
+  hash(): Hash {
+    return this.address().id;
+  }
+
+  entries(): TreeEntry[] {
+    return this.graph
+      .neighborhood(this.nodeAddress, {
+        nodeType: "TREE_ENTRY",
+        edgeType: "INCLUDES",
+        direction: "OUT",
+      })
+      .map(({neighbor}) => new TreeEntry(this.graph, neighbor));
+  }
+
+  entry(name: string): ?TreeEntry {
+    return this.entries().filter((te) => te.name() === name)[0];
+  }
+}
+
+export class TreeEntry extends BaseGitPorcelain {
+  static from(n: BaseGitPorcelain): TreeEntry {
+    if (n.type() !== "TREE_ENTRY") {
+      throw new Error(`Unable to cast ${n.type()} to TreeEntry`);
+    }
+    return new TreeEntry(n.graph, n.nodeAddress);
+  }
+
+  name(): string {
+    const node: Node<TreeEntryNodePayload> = (this.node(): any);
+    return node.payload.name;
+  }
+
+  evolvesTo(): TreeEntry[] {
+    return this.graph
+      .neighborhood(this.nodeAddress, {
+        nodeType: "TREE_ENTRY",
+        edgeType: "BECOMES",
+        direction: "OUT",
+      })
+      .map(({neighbor}) => new TreeEntry(this.graph, neighbor));
+  }
+
+  evolvesFrom(): TreeEntry[] {
+    return this.graph
+      .neighborhood(this.nodeAddress, {
+        nodeType: "TREE_ENTRY",
+        edgeType: "BECOMES",
+        direction: "IN",
+      })
+      .map(({neighbor}) => new TreeEntry(this.graph, neighbor));
+  }
+
+  contents(): Tree | Blob | SubmoduleCommit {
+    const neighbors = this.graph
+      .neighborhood(this.nodeAddress, {
+        edgeType: "HAS_CONTENTS",
+        direction: "OUT",
+      })
+      .map(({neighbor}) => neighbor);
+    if (neighbors.length !== 1) {
+      throw new Error(
+        `Wrong number of contents for ${stringify(this.address())}`
+      );
+    }
+    const n = neighbors[0];
+    const type = n.type;
+    switch (type) {
```

Comment:
IMO, this is easier to read as something like "if the type is BLOB, TREE, or SUBMODULE_COMMIT, then return `asGitPorcelain(n)`, else error." Or, "let `result` be `asGitPorcelain(n)`, and assert that the type of `result` is one of the three valid types."

Reply:
Not a problem now that `asGitPorcelain` has been removed (there's no confusing repetition).